### PR TITLE
23 screen capture and h264 encoding

### PR DESCRIPTION
**Pull request by kousiksatish**

### CODE REVIEW

File: binding.gyp [hunk 1]
```diff
@@ -1,21 +1,35 @@
...
         "native_src/screen_resources.cc",
-        "native_src/x11_util.cc"],
+        "native_src/x11_util.cc",
+        "native_src/output_resolution_with_offset.cc",
+        "native_src/screen_capture_utility.cc",
+        "native_src/encoder.cc",
+        "native_src/base_screen_capturer.cc",
+        "native_src/single_screen_capturer.cc",
+        "native_src/multi_screen_capturer.cc"],
       "defines": [ 'NAPI_DISABLE_CPP_EXCEPTIONS' ]
```
Comment (bavulapati) [inline binding.gyp]: Please comment why are we using this no extension flag

File: src/index.ts [hunk 2]
```diff
@@ -3,9 +3,18 @@ import lockSystem from 'lock-system';
 
 /**
  * DisplayUtility will retrieve the information about the desktop display.
+ * ScreenCaptureUtility will provide functions to capture screen and encode it.
```
Comment (bavulapati) [inline src/index.ts]: It would be better for open source community if we expose the encoding capabilities via different class. Even though we just use it internally.

File: src/index.ts [hunk 3]
```diff
@@ -16,6 +25,12 @@ export interface IResolution {
...
 }
 
+export interface IResolutionWithOffset extends IResolution {
+    offsetX: number;
+    offsetY: number;
+    rrOutput: number;
+}
+
 interface INativeDisplayUtility {
     getConnectedOutputs(): number[] | undefined;
```
Comment (bavulapati) [inline src/index.ts]: Add function comments describing the parameters and mentioning its functionality.
Also for interfaces and classes. Also about return values on positive and negative cases

File: src/test.ts [hunk 4]
```diff
@@ -10,3 +11,52 @@ console.log(displayUtility.getPrimaryRROutput());
 //     width: 1440,
 //     height: 900
 // });
+
+// tslint:disable-next-line:no-console
+console.log(displayUtility.getExtendedMonitorResolution());
```
Comment (bavulapati) [inline src/test.ts]: It would be better if we have unit testing pattern used here. 
We can describe a test case like input, output and we can have an assertion. And then print if it is success or failure.
And at the end we can show the summary like 9/10 tests passed.
or we can use a standard testing framework.
shall we have tests related to each function grouped? May be in separate files.
We can treat this as another pull request if required.

File: native_src/display_utility.cc [hunk 5]
```diff
@@ -72,10 +73,11 @@ Napi::Object GetCurrentResolution(const Napi::CallbackInfo &info)
```
Comment (bavulapati) [inline native_src/display_utility.cc]: In GetOutputName(), let's handle the case when the RROutput is not valid may be by throwing an exception or error. This is not related to your changes

File: native_src/display_utility.cc [hunk 6]
```diff
@@ -72,10 +73,11 @@ Napi::Object GetCurrentResolution(const Napi::CallbackInfo &info)
 
     unsigned int rROutput = info[0].As<Napi::Number>().Int32Value();
```
Comment (bavulapati) [inline native_src/display_utility.cc]: let's handle the case when the RROutput is not valid may be by throwing an exception or error.
There is chance of returning null | undefined. so we may have to update return type of the function.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,99 @@
...
+screenCaptureUtility.init();
+```
+
+### lockUtility.lockScreen(): void
+
+Locks the computer when executed
+
+## Issues
+
+Please feel free to raise issues on this repository, be it bugs or feature requests. We would be happy to address them.
```
Comment (bavulapati) [inline README.md]: I like this addition

File: .vscode/launch.json [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "type": "lldb",
+            "request": "launch",
+            "name": "Debug",
+            "preLaunchTask": "build:dev",
```
[the remark this answers is not in the document]
Comment (kousiksatish) [reply .vscode/launch.json]: Added steps in readme

File: binding.gyp (hunk 1 above)
Comment (kousiksatish) [reply binding.gyp]: Added comments

File: native_src/encoder.cc [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+        {
+            _screenCapturer->InitializeMonitorProperties();
+        }
+    }
+    catch (std::string msg)
+    {
+        throw "ERROR: x264 Encoder initialisation failed." + msg;
+    }
+
+    int width = _screenCapturer->GetWidth();
```
[the remark this answers is not in the document]
Comment (kousiksatish) [reply native_src/encoder.cc]: Done earlier. Please verify

File: native_src/screen_capturer.cc [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+namespace remoting
+{
+ScreenCapturer::ScreenCapturer()
+{
+    _display = XOpenDisplay(NULL);
+    _window = DefaultRootWindow(_display);
+    _multiMonitor = false;
+}
+
+void ScreenCapturer::InitializeMonitorProperties()
```
[the remark this answers is not in the document]
Comment (kousiksatish) [reply native_src/screen_capturer.cc]: Done. Please verify

File: package.json [hunk 11]
```diff
@@ -19,7 +19,8 @@
     "transpile": "tsc --build tsconfig.json",
     "lint": "tslint --project tsconfig.json --config tslint.json --fix",
     "prepare": "npm run transpile",
-    "test": "npm run lint && npm run transpile && node release/test.js"
+    "test": "node-gyp rebuild && npm run lint && npm run transpile && node release/test.js",
+    "build:dev": "npm run transpile && node-gyp rebuild --debug"
   },
   "husky": {
     "hooks": {
```
[the remark this answers is not in the document]
Comment (kousiksatish) [reply package.json]: Done.

File: package.json [hunk 12]
```diff
@@ -50,11 +51,12 @@
...
-    "typescript": "^3.4.5"
+    "@types/node": "^12.0.12",
+    "husky": "^2.7.0",
+    "lint-staged": "^8.2.1",
+    "tslint": "^5.18.0",
+    "tslint-microsoft-contrib": "^6.2.0",
+    "typescript": "^3.5.2"
   },
   "dependencies": {
     "bindings": "^1.5.0",
```
[the remark this answers is not in the document]
Comment (kousiksatish) [reply package.json]: Done.

File: src/index.ts [hunk 13]
```diff
@@ -25,8 +40,15 @@ interface INativeDisplayUtility {
...
+    getExtendedMonitorResolution(): IResolution | undefined;
+    getAllCurrentResolutionsWithOffset(): IResolutionWithOffset[] | undefined;
 }
 
 interface ILockUtility {
     lockScreen(): void;
 }
+
+interface IScreenCaptureUtility {
+    init(rROutput?: number): void;
```
[the remark this answers is not in the document]
Comment (kousiksatish) [reply src/index.ts]: Changed accordingly to have boolean and then rroutput. Please verify

File: src/test.ts (hunk 4 above)
Comment (kousiksatish) [reply src/test.ts]: I would prefer treating it as separate issue and will send another pull request for it

File: src/index.ts (hunk 3 above)
Comment (kousiksatish) [reply src/index.ts]: Added in readme. Will add the same here as comments also.

File: src/index.ts (hunk 2 above)
Comment (kousiksatish) [reply src/index.ts]: I would suggest that we treat this as a separate issue and separate pull request

File: package.json [hunk 14]
```diff
@@ -50,11 +51,12 @@
...
-    "tslint": "^5.16.0",
-    "tslint-microsoft-contrib": "^6.1.1",
-    "typescript": "^3.4.5"
+    "@types/node": "^12.0.12",
+    "husky": "^2.7.0",
+    "lint-staged": "^8.2.1",
+    "tslint": "^5.18.0",
+    "tslint-microsoft-contrib": "^6.2.0",
+    "typescript": "^3.5.2"
   },
```
[the remark this answers is not in the document]
Comment (kousiksatish) [reply package.json]: Added everything except procedure to build x264 library.

File: native_src/encoder.cc [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+    // RGB input information
+    _rgbData = _screenCapturer->GetDataPointer();
+    _rgbPlanes[0] = _rgbData;
+    _rgbPlanes[1] = NULL;
+    _rgbPlanes[2] = NULL;
+    _rgbStride[0] = 4 * width;
+    _rgbStride[1] = 0;
+    _rgbStride[2] = 0;
+
+    // YUV output information
```
[the remark this answers is not in the document]
Comment (kousiksatish) [reply native_src/encoder.cc]: Need more info on this. We will discuss

File: README.md (hunk 7 above)
Comment (kousiksatish) [reply README.md]: (y)

File: native_src/display_utility.cc [hunk 16]
```diff
@@ -72,10 +73,11 @@ Napi::Object GetCurrentResolution(const Napi::CallbackInfo &info)
...
     std::unique_ptr<DisplayUtilityX11> desktopInfo = DisplayUtilityX11::Create();
```
[the remark this answers is not in the document]
Comment (kousiksatish) [reply native_src/display_utility.cc]: As discussed, will handle it in separate issue for next release

File: package.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (kousiksatish) [reply package.json]: Will handle this in separate issue. If it is okay

File: src/index.ts (hunk 3 above)
Comment (kousiksatish) [reply src/index.ts]: Added

File: src/index.ts [hunk 17]
```diff
@@ -25,8 +40,15 @@ interface INativeDisplayUtility {
...
+    getNextFrame(callback: (nextFrame: ArrayBuffer) => void): void;
```
[the remark this answers is not in the document]
Comment (kousiksatish) [reply src/index.ts]: Will handle in separate PR
